fedimint-api/src/multiplexed.rs

```diff
@@ -0,0 +1,180 @@
+#![cfg_attr(target_family = "wasm", allow(unused))]
+use std::fmt::Debug;
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::Arc,
+    time::Duration,
+};
+
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+#[cfg(not(target_family = "wasm"))]
+use tokio::{sync::Mutex, time::sleep};
+#[cfg(target_family = "wasm")]
+type Mutex<T> = std::marker::PhantomData<T>;
+
+use tracing::{debug, error};
+
+use crate::{cancellable::Cancellable, net::peers::AnyPeerConnections, PeerId};
+
+/// TODO: Use proper ModuleId after modularization is complete
+pub type ModuleId = String;
+pub type ModuleIdRef<'a> = &'a str;
+
+/// A `Msg` that can target a specific destination module
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ModuleMultiplexed<Msg> {
+    pub module: String,
+    pub msg: Msg,
+}
+
+struct ModuleMultiplexerOutOfOrder<Msg> {
+    /// Messages per `ModuleId` in a queue each
+    msgs: HashMap<ModuleId, VecDeque<(PeerId, Msg)>>,
+    /// Track pending messages per peer to avoid a potential DoS
+    peer_counts: HashMap<PeerId, u64>,
+}
+
+impl<Msg> Default for ModuleMultiplexerOutOfOrder<Msg> {
+    fn default() -> Self {
+        Self {
+            msgs: Default::default(),
+            peer_counts: Default::default(),
+        }
+    }
+}
+
+/// Shared, mutable (wrapped in mutex) data of [`ModuleMultiplexer`].
+struct ModuleMultiplexerInner<Msg> {
+    /// Underlying connection pool
+    connections: Mutex<AnyPeerConnections<ModuleMultiplexed<Msg>>>,
+    /// Messages that arrived before an interested thread asked for them
+    out_of_order: Mutex<ModuleMultiplexerOutOfOrder<Msg>>,
+}
+
+/// A wrapper around `AnyPeerConnections` multiplexing communication between multiple modules over it
+///
+/// This works by addressing each module when sending, and handling buffering messages received
+/// out of order until they are requested.
+///
+/// This type is thread-safe and can be cheaply cloned.
+#[derive(Clone)]
+#[cfg(not(target_family = "wasm"))]
+pub struct ModuleMultiplexer<Msg> {
+    inner: Arc<ModuleMultiplexerInner<Msg>>,
+}
```

Comment:
This should probably be contextualized a bit. We never want modules to communicate directly, outside HBBFT except for initial DKG where a human is watching to see if it crashes.